jobs/server.go

```diff
@@ -49,22 +52,58 @@ func (srv *JobServer) Config() *model.Config {
 	return srv.ConfigService.Config()
 }
 
-func (srv *JobServer) StartWorkers() {
-	srv.Workers = srv.Workers.Start()
+func (srv *JobServer) StartWorkers() error {
+	srv.mut.Lock()
+	defer srv.mut.Unlock()
+	if srv.workers == nil {
+		return ErrWorkersUninitialized
+	} else if srv.workers.running {
+		return ErrWorkersRunning
```

Comment:
Should this be an error? And the same question for `StartSchedulers`, `StopWorkers`, `StopSchedulers`.

Reply:
It's not necessary but my experience with state handling of this nature is that it's less bug prone to notify the caller about it. We don't want to start the jobs twice so the call becomes a no-op if they are already running. I leave it to the caller to decide whether that's acceptable/expected or not.

Reply:
I'm not following the error case you're thinking of where the caller attempts to start workers that are already running. Do you mind sharing an example?

Reply:
Let's see if I understood:

You are saying that starting an already running process should not be considered an error. 
I am saying it's safer to return an error and let the caller act on it because from the inside I don't know if calling start twice (or more) is the sign of correct functioning and I want to notify the caller that something potentially unexpected has happened. It is then the caller that has the context to figure out if the error can be discarded or logged or else. 

The problem I see with not returning an error is that we might be losing valuable information in case of nasty bugs like races where two calls happen and nothing gets logged because the program think it's all good.

The alternative to all this would be to expose the internal state through some method, then the caller could figure out the state and do the calls accordingly. In that case there wouldn't be a need to return an error for such cases but overall complexity would increase in my view.

Reply:
Yeah, I was questioning whether it's worth breaking the idempotence of the function to use it to test for race conditions. A counter case is:

```bash
touch foo.txt
touch foo.txt
cp foo.txt foo2.txt
cp foo.txt foo2.txt
```

One does not get any "already touched" or "already copied" errors. Adding those errors would create more of a more burden to the caller to capture them appropriately. But those aren't examples of a singleton (like the job server). I guess having that extra logic baked into the function is most appropriate in the case of a singleton.

Totally cool to keep it like this :)

Reply:
Right. If it wasn't written like then we'd expose the need by waiting for someone to ask for a method to check whether the server is running (the file exists in my example). The caller would invoke `(*JobServer).IsRunning`, or whatever.